redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+        connected to the specified node
+        """
+        return ClusterPubSub(self, node=node, host=host, port=port, **kwargs)
+
+    def _determine_nodes(self, *args, **kwargs):
+        command = args[0]
+        node_flag = self.nodes_flags.get(command)
+
+        if node_flag == RANDOM:
+            return [self.get_random_node()]
+        elif node_flag == ALL_PRIMARIES:
+            return self.get_all_primaries()
+        elif node_flag == ALL_REPLICAS:
+            return self.get_all_replicas()
+        elif node_flag == ALL_NODES:
+            return self.get_all_nodes()
+        else:
+            # get the node that holds the key's slot
+            slot = self.determine_slot(*args)
+            return [self.nodes_manager.
+                    get_node_from_slot(slot, self.read_from_replicas
+                                       and command in READ_COMMANDS)]
+
+    def _increment_reinitialize_counter(self, count=1):
+        # In order not to reinitialize the cluster, the user can set
+        # reinitialize_steps to 0.
+        for i in range(min(1, self.reinitialize_steps)):
+            self.reinitialize_counter += count
+            if self.reinitialize_counter % self.reinitialize_steps == 0:
+                self.initialize()
+
+    def _should_reinitialized(self):
+        # In order not to reinitialize the cluster, the user can set
+        # reinitialize_steps to 0.
+        if self.reinitialize_steps == 0:
+            return False
+        else:
+            return self.reinitialize_counter % self.reinitialize_steps == 0
+
+    def keyslot(self, key):
+        """
+        Calculate keyslot for a given key.
+        Tuned for compatibility with python 2.7.x
+        """
+        k = self.encoder.encode(key)
+        return key_slot(k, REDIS_CLUSTER_HASH_SLOTS)
+
+    def determine_slot(self, *args):
+        """
+        figure out what slot based on command and args
+        """
+        if len(args) <= 1:
+            raise RedisClusterException(
+                "No way to dispatch this command to Redis Cluster. "
+                "Missing key."
+            )
+        # @barshaul decide on the implementation of finding key positions in
+        # commands which have a more complicated scheme e.g. 'EVAL', 'EVALSHA',
+        # 'XREADGROUP', 'XREAD', 'XADD'
+
+        command = args[0]
+        key = args[1]
+
+        # OBJECT command uses a special keyword as first positional argument
+        if command == "OBJECT":
+            key = args[2]
+
+        return self.keyslot(key)
+
+    def execute_command(self, *args, **kwargs):
+        """
+        Wrapper for CLUSTERDOWN error handling.
+
+        It will try the number of times specified by the config option
+        "self.cluster_down_retry_attempts" which defaults to 3 unless manually
+        configured.
+
+        If it reaches the number of times, the command will raises
+        ClusterDownException.
+        """
+        if not args:
```

Comment:
IMHO this case really should never occur. We don't check in redis-py elsewhere and assume args[0]

Reply:
removed